### ISSUE TITLE
Support parsing PostgreSQL's dollar quoted string constants

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#6778]
jOOQ should be able to handle PostgreSQL's dollar quoted string constants that work by wrapping a string literal by some `$` enclosed delimiter:

```sql
SELECT $$abc$$, $delimiter$abc$delimiter$
```

This feature should also be made available to the parser!

**Note**: This is difficult to parse, though. In databases like MySQL, `$abc` is a perfectly valid identifier...

----
See:
- https://www.postgresql.org/docs/9.5/static/sql-syntax-lexical.html#SQL-SYNTAX-DOLLAR-QUOTING
- https://stackoverflow.com/q/46731343/521799

Related:
- #6704
- #6777
- #6778

**Comment by timur-sh** [comment]
You mentioned that that is to be difficult to parse. Is there possible to parse a string contains `$$` token only if a `SQLDialect` is known ?

**Comment by lukaseder** [comment]
The problem is that the token isn't really `$$`, but `$[anystringhere]$ ... string content ... $[anystringhere]$`. Ultimately, perhaps it isn't that hard to parse, but I'd like to avoid implementing backtracking in the parser, so a token that starts with `$` may be a string literal or an identifier, depending on whether a second `$` is encountered prior to the first white space. This is only an idea of how this works in PostgreSQL. To be entirely sure, we need the *exact* PostgreSQL grammar or specification for this feature.

**Comment by lukaseder** [comment]
Also, I'd like to avoid making the parsing `SQLDialect` dependent right now. If we had dialect dependent parsers, the expectation would be for the parser to reject any invalid syntax in that dialect, which is much harder to implement than accepting all the merged dialects as we do right now.
